(docs) Staying up to date, troublshooting (#491)
openmrs/openmrs-esm-core: docs/getting_started/troubleshooting.md

# Staying Current & Troubleshooting

> Staying current with the code base is important to make sure we the latest fixes, latest features, and the correct API to develop against. When developing a microfrontend your code will depend on app-shell and core libraries and sometimes other microfrontends. For the OpenMRS project there are several places we should look to make sure our application is kept up to date. If you are having trouble running code locally when it appears to be working on the server ([dev3](https://dev3.openmrs.org/openmrs/spa)), consider taking any or all of these upgrade steps.

Reviewer: ```suggestion
It's important to keep your working code base current with the latest fixes, latest features, and library APIs to develop against. When developing a frontend module your code will depend on the app shell and core libraries, and sometimes other frontend modules. If you are having trouble running code locally when it appears to be working on the server ([dev3](https://dev3.openmrs.org/openmrs/spa)), try the following.

```

Unfortunately we made the decision a while ago to use the word "frontend module" in place of "microfrontend." I maintain that this was a bad idea, but we should be consistent with it. See the comment at the top of the [docs README](https://raw.githubusercontent.com/openmrs/openmrs-esm-core/master/docs/README.md).

> ## Updating core libraries

Reviewer: ```suggestion

## Updating core libraries
```

> If you notice your app suddenly stop working there is a chance that some core library has changed. The main dependencies shared by all OpenMRS microfrontends are  `openmrs` and `@openmrs/esm-framework`.

Reviewer: ```suggestion
If you notice your app suddenly stop working, it might be because a core library has changed in the app shell on the server you are developing against. The main dependencies shared by all OpenMRS microfrontends are  `openmrs` and `@openmrs/esm-framework`. They can be updated like this:
```

```sh
yarn upgrade openmrs @openmrs/esm-framework

> git restore package.json # resets version numbers to "next" 

Reviewer: ```suggestion
git restore package.json  # resets version numbers to "next" 
yarn  # update yarn.lock
```

```

After updating these core libraries be sure to restart your development server

## Updating current branch

> It's good practice to merge the main branch into your working branch often, at least once a week. This way, if anyone else has made changes to the app dependencies or fixed any bugs you will be able to get these benefits. Also, this prevents merge conflicts when trying to make a PR if your files are behind the main branch.

Reviewer: This is basic git knowledge and workflow recommendations. Nothing to do with OpenMRS. The OpenMRS Wiki is already full of this kind of documentation: https://wiki.openmrs.org/display/docs/Using+Git . I would love it if the Frontend docs didn't accumulate this kind of thing.

As a matter of opinion, working branches that go on longer than two weeks are an anti-pattern anyway.

```sh
git checkout main
git pull
git checkout [my-branch-name]
git merge main
```

## Refreshing importmap

> Sometimes the modules being pointed to by the importmap have been updated. If your local development setup is using an old importmap consider replacing the static versions with more recent versions. Another alternative here is to use a dynamic importmap when starting your local server 

Reviewer: ```suggestion
Sometimes the importmap used by `openmrs develop` doesn't have the latest versions of the frontend modules. You can submit a PR to update [that file](https://github.com/openmrs/openmrs-esm-core/blob/master/packages/shell/esm-app-shell/src/assets/importmap.json), or you can use an import map from the internet like so:
```

This is good and helpful documentation. The fact that we have to include it, including the suggestion to update the stale file, suggests that this is an especially bad piece of our DX. We should consider either
1. Going back to just using `yarn serve` with import-map-overrides to load modules into the app OR
2. Making `openmrs develop` use the remote importmap by default and getting rid of this maintenance burden of an `importmap.json`

```sh

> # an example of pointing a local development server to a dynamic importmap

Reviewer: ```suggestion
```

yarn start --importmap "https://spa-modules.nyc3.digitaloceanspaces.com/import-map.json"
```

## Clearing out the browser cache

> By default the module javascript files are cached by your browser for 30 days. This is thanks to a [configuration setting](https://github.com/openmrs/openmrs-contrib-ansible-docker-compose/blob/c36115ca22b8fb842f8cf0ff745d1d35a4567912/files/emr-3-dev/proxy.conf#L97) on the server. This is a fine setting for end users, and caching is useful to make sure we don't need to download the entire app again with each reload, but as developers we often need the latest version of a core module or microfrontend as soon as it comes out. The easiest way to trigger a re-download of the app javascript files is to open your browser's implementer tools, going to the "Network" tab and checking "Disable Cache". Then, force reload the page. You should get brand new javascript files from the server. After you have refreshed the page to replace your cached script files, you can uncheck the "Disable Cache" option to improve the application behavior and not re-download script files on every reload. 

Reviewer: ```suggestion
By default the server [tells your browser](https://github.com/openmrs/openmrs-contrib-ansible-docker-compose/blob/c36115ca22b8fb842f8cf0ff745d1d35a4567912/files/emr-3-dev/proxy.conf#L97) to cache JavaScript files for 30 days. You can [bypass your browser's cache](https://en.wikipedia.org/wiki/Wikipedia:Bypass_your_cache) when refreshing the page. If you want to keep the cache disabled while you are developing, open your browser's developer tools, go to the "Network" tab, and check "Disable Cache." The cache will only be disabled while the developer tools are open, and only for the page that they are attached to.
```

How to open browser developer tools (chrome)
<img width="933" alt="open-devtools-chrome" src="https://user-images.githubusercontent.com/5445264/182458545-ba701bb0-1cfe-4083-98bb-5e7338d97065.png">

How to disable asset (javascript, css) cache (chrome)
<img width="936" alt="disable-cache-chrome" src="https://user-images.githubusercontent.com/5445264/182458626-bab053e1-0d4c-4d71-a69a-1357fc1dd13e.png">

## Restarting your computer

> Its possible for browser caches to not get cleared properly, or else for applications to be left in an inconsistent state which causes them to run poorly, slowly, or not at all. Sometimes these issues are only recoverable with restarting your computer

Reviewer: I've never had to do this one while doing frontend development :shrug: 

In this vein, I have occasionally had to delete `node_modules` and re-run `yarn` though.

```suggestion
Its possible for browser caches to not get cleared properly, or else for applications to be left in an inconsistent state which causes them to run poorly, slowly, or not at all. Sometimes these issues are only recoverable with restarting your computer.
```
